server/src/main/java/org/elasticsearch/common/network/HandlingTimeTracker.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.network;
+
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * Tracks how long message handling takes on a transport thread as a histogram with fixed buckets.
+ */
+public class HandlingTimeTracker {
+
+    public static int[] getBucketUpperBounds() {
+        // Default clock resolution is 200ms so we have buckets for the 0-tick and 1-tick cases, then go up in powers of two
```

Comment:
Should we use the raw time rather than the cached time so that we get better resolution?

Reply:
Hmm we'd have to if we actually want fine grained resolution. I'd say it's ok, having fine grained information here is quite valuable/interesting and relative to the cost of reading from a channel etc, `System.nanoTime` should be trivial.